client/src/partials/budget/analysis/analysis_budget.js

```diff
@@ -0,0 +1,436 @@
+          dependencies.previousFYBudget.query.where = ['period.fiscal_year_id=' + fiscal_year_id];
+        } else {
+          dependencies.previousFYBudget.query.where = ['period.fiscal_year_id=' + fiscal_year_id, 'AND', 'period.period_number='+selectedPeriod.period_number];
+        }
+        validate.refresh(dependencies, ['previousFYBudget'])
+        .then(function (model) {
+          var data = model.previousFYBudget.data;
+          handleFiscalYearBudget(fiscal_year_id, data);
+        });
+      }
+
+      function handleFiscalYearBudget(fiscal_year_id, data) {
+        $scope.fiscalYearBudget[fiscal_year_id] = data;
+      }
+    }
+
+    function loadFiscalYears(models) {
+      angular.extend($scope, models);
+    }
+
+    appstate.register('enterprise', function (enterprise) {
+      enterprise_id = Number(enterprise.id);
+      $scope.enterprise = enterprise;
+      dependencies.fiscal_years.query.where = [ 'fiscal_year.enterprise_id=' + enterprise_id ];
+      validate.process(dependencies, ['fiscal_years'])
+        .then(loadFiscalYears);
+    });
+
+    function loadPeriod(fiscal_year_id) {
+      dependencies.period = {
+        query : {
+          tables : {
+            'period' : { columns : ['id', 'period_number', 'period_start', 'period_stop'] }
+          },
+          where : ['period.fiscal_year_id=' + fiscal_year_id, 'AND', 'period.id<>0']
+        }
+      };
+
+      validate.refresh(dependencies, ['period'])
+      .then(function (model) {
+        session.periods = model.period.data;
+        loadPreviousFiscalYears(fiscal_year_id);
+      });
+    }
+
+    function loadPreviousFiscalYears(fiscal_year_id) {
+      session.previous_fiscal_years = [];
+      var current = $scope.fiscal_years.get(fiscal_year_id);
+      if (current.previous_fiscal_year) {
+        session.previous_fiscal_years = $scope.fiscal_years.data.filter(function (fy) {
+          var currentDate = new Date(current.start_year, current.start_month, 1);
+          var otherDate = new Date(fy.start_year, fy.start_month, 1);
+          return (currentDate > otherDate) ? true : false;
+        });
+      }
+    }
+
+    function budgetAnalysis() {
+
+      if (config.fiscal_year_id && config.period_id) {
+        var selectedPeriod = session.periods.filter(function (p) {
+          return p.id === config.period_id;
+        })[0];
+        session.selectedPeriod = $translate.instant($scope.months[selectedPeriod.period_number]);
+        session.selectedFiscalYear = $scope.fiscal_years.get(config.fiscal_year_id);
+        getSelectPreviousFY()
+        .then(previousFYBudget)
+        .then(displayAccounts);
+      }
+
+      function getSelectPreviousFY() {
+        var def = $q.defer();
+        session.selectedPreviousFY = session.previous_fiscal_years.filter(function(fy) {
+          return fy.checked;
+        });
+        def.resolve(session.selectedPreviousFY);
+        return def.promise;
+      }
+    }
+
+    function exportToCSV() {
```

Comment:
This is a good beginning.  Could you make this into a service?  The service should take in an array of JSON objects and a configuration object, and return a downloadable CSV file.  For example, here is how I would use it:
```
var data = [{ id: 1, name : 'Jaquline'}, {id:2, name: 'Bombo'}, { id : 4, name  : 'Yvete' }];

// trigger a CSV download using a service
CSVService.download(data, { strict : true, headers: { id : 'ID', name : 'Patient\'s Last Name'}});
```

The CSVService should first check that each JSON object in the array has all the same properties/keys.  For example, it should accept `[{ id :1 }, { id: 2}]`, but reject `[{id : 1}, {id :3, name : 'Frank'}]` and `[{x : 1} , {y:2}]`.

If any values are null, they should be empty strings.  If you pass the config option `strict` to the CSVService, it should throw an error if you are missing values.

Finally, the config option `headers` should give an object that describes what the header row of the CSV should be.